Add @azure/instrumentation-opentelemetry package

maorleger opened the pull request:
## What

- Adds a new package @azure/instrumentation-opentelemetry to the SDK
- Implements the new core-tracing interfaces for OpenTelemetry

## Why

As core-tracing no longer imports @opentelemetry/api, the tracer / instrumenter will be a simple no-op (see #18725 for reference).

For users who want to enable OpenTelemetry based instrumentation, this package would allow them to plug it into the OpenTelemetry ecosystem and provide an Azure implementation of an OpenTelemetry tracer.

## Example usage

```ts
// Everything here is vanilla OTel except the `createAzureInstrumentation` call
const provider = new NodeTracerProvider();
provider.addSpanProcessor(new SimpleSpanProcessor(new JaegerExporter()));
registerInstrumentations({
  instrumentations: [createAzureInstrumentation()],
});
provider.register();
```

Resolves #18651 

## Callouts

One consideration is that this isn't quite finished, but since we're merging it into a feature branch it seems ready to enable. I tested this in Node and the browser and all seems ok 👍 

Node, using Jaeger, with both Azure SDK and Http instrumentation enabled:

![image](https://user-images.githubusercontent.com/753570/142497587-d243139c-0667-4d62-803a-fd99202577cc.png)

Browser, using Console exporter

![image](https://user-images.githubusercontent.com/753570/142497721-7a685a29-cc79-4f99-b9bf-9f3368a1bcbd.png)

azure-sdk commented:
API changes have been detected in `@azure/core-auth`. You can review API changes [here](https://apiview.dev/Assemblies/Review/b0bd88f50fa24faaadbf2297c15730ca?diffOnly=True&diffRevisionId=1809779bf76043988e80441bab2e594d)

azure-sdk commented:
API changes have been detected in `@azure/core-tracing`. You can review API changes [here](https://apiview.dev/Assemblies/Review/6e5c0e25ac0b4f718169d3df61fed963?diffOnly=True&diffRevisionId=f4f62ac9077d42c48d29e112284bf2a7)

xirzec commented on sdk/instrumentation/instrumentation-opentelemetry/review/instrumentation-opentelemetry.api.md:
```diff
@@ -0,0 +1,71 @@
+import { TracingSpanContext } from '@azure/core-tracing';
+
+// @public
+export class AzureSDKInstrumentation extends InstrumentationBase {
```
any chance we can get away with not having this be a class and also not doing cross-package inheritance? The latter has burned us a few times and the former is more because I resist classes whenever we're not creating a service client. 😄

xirzec replied:
Belatedly realized this was coming from OT, agreed we might not be able to avoid it

maorleger replied:
Ended up using Jeff's suggestion and this is now completely hidden from the public API using `createAzureInstrumentation` instead...

maorleger commented on sdk/core/core-tracing/src/useInstrumenter.browser.ts:
```diff
@@ -0,0 +1,16 @@
+import { Instrumenter } from "./interfaces";
+
+/** @internal */
+globalThis.instrumenterImplementation = new NoOpInstrumenter();
```
In the browser we set it on the global object to have a single instrumenter to set.

In Node, the instrumentation package will patch any `require` of core-tracing and set the instrumenter at the module level.

Let me know if this is not the way to go...

xirzec replied:
it feels weird to expose the noop instrumenter if none is set yet... I wonder if we could do something for the browser like "if a global instrumenter is set, use that, otherwise us the one we've cached with useInstrumenter" - that way the global variable functions more like an environment variable override

maorleger commented on sdk/core/core-tracing/src/useInstrumenter.ts:
```diff
@@ -0,0 +1,14 @@
+ * @param instrumenter - The instrumenter implementation to use.
+ */
+export function useInstrumenter(instrumenter: Instrumenter): void {
+  instrumenterImplementation = instrumenter;
```
In node the OTel instrumentation library handles patching any imports of @azure/core-tracing so we can ensure we always call useInstrumenter...

maorleger commented on sdk/core/core-tracing/src/globals.d.ts:
```diff
@@ -0,0 +1,10 @@
+
+// eslint-disable-next-line @azure/azure-sdk/ts-no-namespaces
+declare global {
+  // eslint-disable-next-line no-var
```
Really not sure about this - would love any suggestions such that I can do `globalThis.instrumenterImplementation = something`

xirzec replied:
I think using `Symbol.for()` and using the resulting symbol as a property name will work better than using some normal identifier

xirzec reviewed (commented):
A couple quick thoughts! Liking the energy here

maorleger commented:
I think I addressed the feedback @xirzec - mind giving this another look?

@lmolkova would love your review as well from the tracing side (see description for usage example)

lmolkova commented on common/config/rush/pnpm-lock.yaml:
```diff
@@ -1509,6 +1511,11 @@ packages:
       '@opentelemetry/api': 1.0.3
     dev: false
 
+  /@opentelemetry/api-metrics/0.27.0:
```
is this file relevant in case of user apps? if so, can it be removed for now? same for instrumentation-api package below

maorleger replied:
This looks like a transitive dependency pulled in by @opentelemetry/instrumentation 

```bash
instrumentation-opentelemetry instrumentation-opentelemetry % npm ls @opentelemetry/api-metrics               
@azure/instrumentation-opentelemetry@1.0.0-beta.1 /home/mleger/workspace/worktrees/tracing/sdk/instrumentation/instrumentation-opentelemetry
└─┬ @opentelemetry/instrumentation@0.27.0 -> /home/mleger/workspace/worktrees/tracing/common/temp/node_modules/.pnpm/@opentelemetry+instrumentation@0.27.0_@opentelemetry+api@1.0.3/node_modules/@opentelemetry/instrumentation
  └── UNMET DEPENDENCY @opentelemetry/api-metrics@0.27.0

npm ERR! missing: @opentelemetry/api-metrics@0.27.0, required by @opentelemetry/instrumentation@0.27.0
```

So that explains why it shows up in lock-file. Unfortunately I don't know if we can yank it out

lmolkova reviewed (approved):
Looks great!

xirzec commented on sdk/instrumentation/instrumentation-opentelemetry/src/spanWrapper.ts:
```diff
@@ -0,0 +1,54 @@
+   *
+   * @returns The underlying span
+   */
+  unwrap(): unknown {
```
can't we type it here, since we have the real type anyway? It should still be compatible since the interface it implements has a broader return type

```suggestion
  unwrap(): Span {
```

xirzec commented on sdk/instrumentation/instrumentation-opentelemetry/src/transformations.ts:
```diff
@@ -0,0 +1,192 @@
+ */
+export function toOpenTelemetrySpanKind<K extends TracingSpanKind>(
+  tracingSpanKind?: K
+): SpanKindMapping[K] {
```
some fancy TS in here 😄

maorleger replied:
You probably already figured this out, but this is courtesy of Will - it's way above my paygrade 😄

xirzec commented on sdk/instrumentation/instrumentation-opentelemetry/test/public/util/testTracer.ts:
```diff
@@ -0,0 +1,170 @@
+// Copyright (c) Microsoft Corporation.
```
I'm assuming we'd update the test-utils package after we finish abstracting OpenTelemetry out of the main repo? 

Would there be an equivalent set of "test" implementations for Instrumentation that we replace our existing TestSpan/TestTracer with?

maorleger replied:
That's the current plan, yup! And once we are happy with test-utils we can go back and reuse those here (unless we end up with OTel specific tests we'll see...)

xirzec commented on sdk/core/core-tracing/src/useInstrumenter.browser.ts:
```diff
@@ -0,0 +1,41 @@
+import { NoOpInstrumenter } from "./instrumenter";
+import { Instrumenter } from "./interfaces";
+
+const GLOBAL_INSTRUMENTER_SYMBOL = Symbol("@azure/core-tracing instrumenter");
+
+interface Cache {
+  instrumenter: Instrumenter;
+}
+
+let cache: Cache;
+function getCache(): Cache {
+  if (!cache) {
+    const globalObj: any = globalThis;
+    if (!globalObj[GLOBAL_INSTRUMENTER_SYMBOL]) {
+      globalObj[GLOBAL_INSTRUMENTER_SYMBOL] = {
+        instrumenter: new NoOpInstrumenter()
+      };
+    }
+    cache = globalObj[GLOBAL_INSTRUMENTER_SYMBOL];
+  }
+  return cache;
+}
```
do we need any of this? It seems like the browser version just hooks `.enable()` now instead of placing the Instrumenter on the global object?

maorleger replied:
My original thought here was to support side-by-side installs of core-tracing; however, as discussed offline going to punt it as YAGNI and make an issue to just ensure that scenario works as expected 👍

maorleger replied:
Ok, I removed the browser-specific global crud and created https://github.com/Azure/azure-sdk-for-js/issues/18866 to track that we ensure side-by-side works as expected (or at least have a conversation about it)

xirzec reviewed (commented):
A few last remarks, but this is looking really polished!

azure-sdk commented:
API changes have been detected in `@azure-rest/core-client`. You can review API changes [here](https://apiview.dev/Assemblies/Review/d82d81d526b44f8d97500ecd30c35c55?diffOnly=True&diffRevisionId=ed46ec768eeb44119d885331ce9f392a)

**API changes**
```diff
-     binaryResponse?: boolean;
```

azure-sdk commented:
API changes have been detected in `@azure/core-http`. You can review API changes [here](https://apiview.dev/Assemblies/Review/11abe967cd1c40d3a3abe3719583551b?diffOnly=True&diffRevisionId=0bc99905aa0f4114bccba8a7d274d49b)

azure-sdk commented:
API changes have been detected in `@azure/core-rest-pipeline`. You can review API changes [here](https://apiview.dev/Assemblies/Review/66d8b69f4ede4d05a7af68c12fc32695?diffOnly=True&diffRevisionId=65e9418871f0422182c4b6cd59b684ba)

**API changes**
```diff
-     logger?: AzureLogger;
-     logger?: AzureLogger;
-     logger?: AzureLogger;
```

azure-sdk commented:
API changes have been detected in `@azure/core-tracing`. You can review API changes [here](https://apiview.dev/Assemblies/Review/6e5c0e25ac0b4f718169d3df61fed963?diffOnly=True&diffRevisionId=f4f62ac9077d42c48d29e112284bf2a7)

xirzec commented:
@praveenkuttappan looks like the API diff'er struggles a bit when computing diffs for packages not touched by the current PR. Maybe because this fork needs to be synced with master?

xirzec reviewed (approved):
Latest changes look good
